ruler/ruler.py

```diff
@@ -1,177 +1,265 @@
 import numpy as np
+from typing import Iterable, Optional, Tuple
 
-solid_indicator = ['solid','solids','black','s','b','true','1',True,1]
-void_indicator = ['invert','void','white','v','w','false','0',True,0]
-minimum_indicator = ['minimum','minimal','min']
-both_indicator = ['both','two','2',2]
-
-def ruler_initialize(arr,phys_size,margin_size):
-    if isinstance(phys_size,np.ndarray):
-        phys_size = phys_size[np.squeeze(phys_size).nonzero()] # physical size
-    elif isinstance(phys_size,list) or isinstance(phys_size,tuple):
-        phys_size = np.array(phys_size)[np.squeeze(np.array(phys_size)).nonzero()]
-    elif isinstance(phys_size,float) or isinstance(phys_size,int):
-        phys_size = np.array([phys_size])
-    else:
-        AssertionError("Invalid format of the physical size.")
+solid_indicator = ['solid', 'solids', 'black', 's', 'b', 'true', '1', True, 1]
+void_indicator = ['invert', 'void', 'white', 'v', 'w', 'false', '0', True, 0]
+minimum_indicator = ['minimum', 'minimal', 'min']
+both_indicator = ['both', 'two', '2', 2]
 
-    dims = len(phys_size) # dimension, should be 1, 2 or 3
-    assert np.squeeze(arr).ndim == dims, 'The physical size and the dimension of the input array do not match.'
 
+def minimum_length(
+        arr: Iterable[np.ndarray],
+        phys_size: Iterable[np.ndarray],
+        margin_size: Iterable[np.ndarray] = np.zeros((2, 2)),
+        measure_region: str = 'min',
+        pad_mode: Optional[str] = None,
+        with_boundary: bool = False) -> Tuple[float, Optional[float]]:
     '''
-    Users sometimes need to discard certain marginal regions, the withds of which are specified by margin_size.
-    The first pair of elements correspond to the first dimension, the second pair of elements correspond to the second dimension,
-    and the third pair of elements, if there are, correspond to the third dimension.
-    '''
-    margin_size = np.reshape(margin_size,(-1,2))
-    if len(margin_size[:,0])>dims:
-        margin_size = margin_size[0:dims,:]
-        print('Warning: The row number of margin_size is larger than the dimension; The redundant row(s) will be discarded.')
-    return phys_size, margin_size
-
-
-def minimum_length(arr,phys_size,margin_size=np.zeros((2,2)),measure_region='solid',pad_mode=None):
+    Compute the minimum length scale in a design pattern.
+
+    arr: input array that represents a design pattern.
```

Comment:
These first three parameters are typed as `Iterable` and it's not clear from these docstrings how that is supposed to be interpreted. Is one supposed to provide a list of designs and the function will return a list of measured feature sizes (per provided design)?

Reply:
> Is one supposed to provide a list of designs and the function will return a list of measured feature sizes (per provided design)?

No, the user is supposed to provide one design every time instead of a list of designs. As mentioned in the annotation, `arr: input array that represents a design pattern.` Do you think it is still not clear?

Reply:
Then `Iterable[np.ndarray]` is not the correct type annotation. It should be `np.ndarray`.

Reply:
I see. Thanks!